docs/auth/backend-auth.md

```diff
@@ -0,0 +1,82 @@
+---
+
+##
+
+By default, the backstage backend does not require authentication. Securing your backend requires additional configuration or a separate service like Google's Identity-Aware Proxy. Since the backend is served as an express app we can leverage express middleware to require JWT bearer tokens for all requests. The `@backstage/plugin-auth-node` package provides several factory methods for creating express middleware that can be used to secure your backend:
+
+- `createRequireJwtTokenMiddleware`: Creates middleware that requires a JWT bearer token for all requests.
+- `createRequireUserTokenMiddleware`: Creates middleware that requires a JWT bearer token for all requests and verifies that the token is a valid user token.
+- `createRequireServerTokenMiddleware`: Creates middleware a JWT bearer token for all requests and verifies that the token is a valid server token (see [service to service Auth](./service-to-service-auth.md)).
+
+> NOTE: `createRequireJwtTokenMiddleware` does not validate the JWT tokens, instead you must specify an express middleware that will be invoked when the token has the specified audience. `createRequireUserTokenMiddleware` and `createRequireServerTokenMiddleware` both validate the tokens.
+
+The middleware created by these factories only verifies that a valid JWT token is provided and thus adds **authentication** to the api endpoint. The middleware cannot check whether the identity enclosed in the token has access to specific resources (**authorisation**). Each plugin backend is responsible for checking the permissions of the identity in the token if required.
+
+To configure all backend routes, except `/auth`, to require a valid user or server token, you can use the following code:
+
+```ts title="packages/backend/src/index.ts"
+// ...
+/* highlight-add-start */
+import {
+  createRequireJwtTokenMiddleware,
+  createRequireServerTokenMiddleware,
+  createRequireUserTokenMiddleware,
+} from '@internal/plugin-auth-node';
+/* highlight-add-end */
+
+// ...
+async function main() {
+  // ...
+  const createEnv = makeCreateEnv(config);
+
+  /* highlight-add-start */
+  const tokenManager = ServerTokenManager.fromConfig(config, {
+    logger: getRootLogger(),
+  });
+  const discovery = SingleHostDiscovery.fromConfig(config);
+  const identity = DefaultIdentityClient.create({
+    discovery,
+  });
+
+  const requireBackstageToken = createRequireJwtTokenMiddleware({
+    backstage: createRequireUserTokenMiddleware(identity),
+    'backstage-server': createRequireServerTokenMiddleware(tokenManager),
+  });
+  /* highlight-add-end */
+
+  const apiRouter = Router();
+
+  /* highlight-add-next-line */
+  // unauthenticated routes
+  apiRouter.use('/auth', await auth(authEnv));
+
+  /* highlight-add-next-line */
+  // authenticated routes
+  apiRouter.use(
+    '/catalog',
+    /* highlight-add-start */ requireBackstageToken,
+    /* highlight-add-end */ await catalog(catalogEnv),
+  );
```

Comment:
We're about to move away from this way of wiring up the backend, see the [new backend system](https://backstage.io/docs/backend-system/). The app doesn't own the plugin routing in that system, it's instead done by plugins. What I'm feeling makes the most sense is to make it an opt-out for plugins to not require auth on certain routes. Exactly how we do that is something to be figured out 😅

Either way, I'm feeling that these middleware is something we'll end up needing to build into the new system rather than make available for plugins. I don't think we should ship these as helpers for the old system from `auth-node` either, it's too much work to push out for something that should go away.

Reply:
I wasn't aware of that. The reason I added these factory functions to `auth-node` was to help others setup authentication using express middleware. But if the backend is going to replaced with `vnext` soon. If you are suggesting to not merge this PR then I'll close the PR.

Reply:
Regarding how to do the opt-out on certain routes, could [express-unless](https://www.npmjs.com/package/express-unless) work? You'd have to expose the auth middleware to the plugin. Or something similar to [express-unless](https://www.npmjs.com/package/express-unless) that would allow you to configure the auth middleware without exposing the actual middleware/handler. In that case so you can do something like this:
```typescript
export const plugin = createBackendPlugin({
  pluginId: 'plugin',
  register(env) {
    // ...
    async init({ http, requireAuthentication}) {
      requireAuthentication.unless({
       path: /^.+\.css/,
      });

      http.use(
        await createRouter(),
      );
    },
});
```

Reply:
Would be sweet if it could be done as a middleware on the router, but that might be pretty tricky. Doing it based on patterns within the plugin is unnecessarily complex imo

Reply:
Was just thinking out loud on how it could be done without changing the `createRouter()` function. There are ways to filter with [express-unless](https://www.npmjs.com/package/express-unless) but you have a point that it makes less intuitive. 

Maybe you could have `http.use` configure authentication by default but have an overload that passes the auth middleware to the plugin instead of configuring it by default. 

- Configure auth by default:
```typescript
export const plugin = createBackendPlugin({
  pluginId: 'plugin',
  register(env) {
    // ...
    async init({ http})
      // requires authentication by default
      http.use(
        await createRouter(),
      );
    },
});
```
- Pass the auth middleware to the plugin so it can invoke `router.get('path', requireAuthentication, handler)`
```typescript
export const plugin = createBackendPlugin({
  pluginId: 'plugin',
  register(env) {
    // ...
    async init({ http })
      // requires no authentication by default
      http.use(requireAuthentication => 
        await createRouter(requireAuthentication),
      );
    },
});

const createRouter = (requireAuthentication: Handler) => {
   router.get('/v1/todo', requireAuthentication, (req, res) => { ... }
   router.get('/v1/public', (req, res) => { ... }
}
```